src/psyclone/psyir/nodes/omp_directives.py

```diff
@@ -638,6 +705,946 @@ def _encloses_omp_directive(self):
+                                                   parent=parent)
+        # We don't know if we have a parent OMPParallelClause at initialisation
+        # so we can only create dummy clauses for now.
+        self.children.append(OMPPrivateClause())
+        self.children.append(OMPFirstprivateClause())
+        self.children.append(OMPSharedClause())
+        self.children.append(OMPDependClause(
+                depend_type=OMPDependClause.DependClauseTypes.IN))
+        self.children.append(OMPDependClause(
+                depend_type=OMPDependClause.DependClauseTypes.OUT))
+        # We store the symbol names for the parent loops so we can work out the
+        # "chunked" loop variables.
+        self._parent_loop_vars = []
+        self._parent_loops = []
+        self._proxy_loop_vars = {}
+        self._parent_parallel = None
+        self._parallel_private = None
+
+    @staticmethod
+    def _validate_child(position, child):
+        '''
+         Decides whether a given child and position are valid for this node.
+         The rules are:
+         1. Child 0 must always be a Schedule.
+         2. Child 1 must always be an OMPPrivateClause
+         3. Child 2 must always be an OMPFirstprivateClause
+         4. Child 3 must always be an OMPSharedClause
+         5. Child 4 and 5 must always be OMPDependClauses
+
+        :param int position: the position to be validated.
+        :param child: a child to be validated.
+        :type child: :py:class:`psyclone.psyir.nodes.Node`
+
+        :return: whether the given child and position are valid for this node.
+        :rtype: bool
+
+        '''
+        if position == 0:
+            return isinstance(child, Schedule)
+        if position == 1:
+            return isinstance(child, OMPPrivateClause)
+        if position == 2:
+            return isinstance(child, OMPFirstprivateClause)
+        if position == 3:
+            return isinstance(child, OMPSharedClause)
+        if position == 4 or position == 5:
+            return isinstance(child, OMPDependClause)
+        return False
+
+    def validate_global_constraints(self):
+        '''
+        Perform validation checks that can only be done at code-generation
+        time.
+
+        :raises GenerationError: if this OMPTaskDirective is not \
+                                 enclosed within an OpenMP serial region.
+        '''
+        # It is only at the point of code generation that we can check for
+        # correctness (given that we don't mandate the order that a user
+        # can apply transformations to the code). A taskloop
+        # directive, we must have an OMPSerialDirective as an
+        # ancestor back up the tree.
+        if not self.ancestor(OMPSerialDirective):
+            raise GenerationError(
+                "OMPTaskDirective must be inside an OMP Serial region "
+                "but could not find an ancestor node.")
+
+    def _find_parent_loop_vars(self):
```

Comment:
I thought it might be potentially called quite frequently and could be quite heavy so I thought just call it once and store the solution, but I think we can make a final decision on that later.